File: 0501-1000/0530-Minimum-Absolute-Difference-in-BST/java-0530/src/Solution3.java
/// Source : https://leetcode.com/problems/minimum-absolute-difference-in-bst/description/
/// Author : liuyubobobo
/// Time   : 2018-04-30

import java.util.ArrayList;
import java.util.Collections;

/// Using inorder to traverse all nodes
/// and calculates the result during the inorder traverse
///
/// Time Complexity: O(n)
/// Space Complexity: O(logn)
class Solution3 {

    private Integer prev = null;

    public int getMinimumDifference(TreeNode root) {

        prev = null;
        return inOrder(root);
    }

    private int inOrder(TreeNode node){

        if(node == null)
            return Integer.MAX_VALUE;

        int res = Integer.MAX_VALUE;

        res = Math.min(res, inOrder(node.left));

        if(prev != null)
            res = Math.min(res, node.val - prev);
        prev = node.val;

        res = Math.min(res, inOrder(node.right));

        return res;
    }
}

File: 0001-0500/0064-Minimum-Path-Sum/py-0064/Solution3.py
# Source : https://leetcode.com/problems/minimum-path-sum/
# Author : penpenps
# Time   : 2019-07-10

from typing import List

# Use single list to keep accumulate dp result
# Time Complexity: O(n*m)
# Space Complexity: O(n)
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        if not grid or not grid[0]: 
            return 0
        n = len(grid)
        m = len(grid[0])
        dp = [0]*m
        dp[0] = grid[0][0]
        for i in range(1, m):
            dp[i] = dp[i-1] + grid[0][i]
        
        for i in range(1, n):
            dp[0] += grid[i][0]
            for j in range(1, m):
                dp[j] = min(dp[j-1], dp[j]) + grid[i][j]
        
        return dp[m-1]

if __name__ == '__main__':
    solution = Solution()
    grid = [
    [1,3,1],
    [1,5,1],
    [4,2,1]
    ]
    print(solution.minPathSum(grid))